### PR TITLE
Initial versions of StringHelper.Join

### PR DESCRIPTION
**Pull request by frankbuckley**
There is more we can do to optimise but taking this now to unblock me downstream.

Part of #79 

cc. @adambuckley12 , @rabuckley

### CODE REVIEW

File: src/DSE.Open/Text/StringConcatenator.cs
Comment (rabuckley) [inline src/DSE.Open/Text/StringConcatenator.cs:1]: Is this staying?

Comment (frankbuckley) [reply src/DSE.Open/Text/StringConcatenator.cs:1]: For now as it is in benchmark - will remove at some point

File: src/DSE.Open/Text/StringHelper.Join.cs [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+            {
+                return string.Empty;
+            }
+
+            if (length <= StackallocThresholds.MaxCharLength)
+            {
+                return JoinPreAllocated(separator.AsSpan(), collection, finalSeparator.AsSpan(), length);
+            }
+
+            return string.Create(length, collection, (chars, state) =>
```
Comment (rabuckley) [inline src/DSE.Open/Text/StringHelper.Join.cs:72]: This should take `count` in the state, otherwise it's going to allocate a new delegate.

Comment (frankbuckley) [reply src/DSE.Open/Text/StringHelper.Join.cs:72]: Good catch 👍

**Review by rabuckley — APPROVED** [review]

**Pull request merged**